compiler/src/main/java/org/qbicc/type/definition/LoadedTypeDefinitionImpl.java

```diff
@@ -115,13 +118,25 @@ public LoadedTypeDefinition getInterface(final int index) throws IndexOutOfBound
     }
 
     public void forEachInterfaceFullImplementedSet(Consumer<LoadedTypeDefinition> function) {
-        for (LoadedTypeDefinition i : interfaces) {
-            function.accept(i);
+        ArrayDeque<LoadedTypeDefinition> worklist = new ArrayDeque<>();
+        LoadedTypeDefinition cur = this;
+        while (cur != null) {
+            Collections.addAll(worklist, cur.getInterfaces());
+            cur = cur.getSuperClass();
         }
-        // Walk up the heirarchy and visit each inteface from the the superclass
-        LoadedTypeDefinition superClass = getSuperClass();
-        if (superClass != null) {
-            superClass.forEachInterfaceFullImplementedSet(function);
+        if (this.isInterface()) {
+            worklist.add(this);
+        }
+        // worklist contains all directly implemented interfaces in the super class chain
+        HashSet<LoadedTypeDefinition> visited = new HashSet<>();
+        while (!worklist.isEmpty()) {
+            LoadedTypeDefinition ltd = worklist.pop();
+            if (!visited.contains(ltd)) {
+                visited.add(ltd);
```

Comment:
This can be merged into:
```suggestion
            if (visited.add(ltd)) {
```